mlflow/transformers.py

```diff
@@ -0,0 +1,851 @@
+    conf = {
+        "task": flavor_config[_TASK_KEY],
+        "model": model_instance.from_pretrained(pipeline_path),
+    }
+
+    if flavor_config.get(_PROCESSOR_TYPE_KEY, None):
+        conf[_PROCESSOR_KEY] = _load_component(
+            local_path, _PROCESSOR_KEY, flavor_config[_PROCESSOR_TYPE_KEY]
+        )
+
+    for component_key in flavor_config[_COMPONENTS_BINARY_KEY]:
+        component_type_key = f"{component_key}_type"
+        component_type = flavor_config[component_type_key]
+        conf[component_key] = _load_component(local_path, component_key, component_type)
+    return conf
+
+
+def _fetch_model_card(model_or_pipeline):
+    """
+    Attempts to retrieve the model card for the specified model architecture iff the
+    `huggingface_hub` library is installed. If a card cannot be found in the registry or
+    the library is not installed, returns None.
+    """
+    with contextlib.suppress(ImportError):
+        from transformers import Pipeline
+
+        hub = importlib.import_module("huggingface_hub")
+        model = (
+            model_or_pipeline.model
+            if isinstance(model_or_pipeline, Pipeline)
+            else model_or_pipeline[_MODEL_KEY]
+        )
+        return hub.ModelCard.load(model.name_or_path)
+
+
+def _build_pipeline_from_model_input(model, task: str):
+    """
+    Utility for generating a pipeline from component parts. If required components are not
+    specified, use the transformers library pipeline component validation to force raising an
+    exception. The underlying Exception thrown in transformers is verbose enough for diagnosis.
+    """
+    from transformers import pipeline
+
+    pipeline_config = {
+        "task": task,
+        **model,
+    }
+    try:
+        return pipeline(**pipeline_config)
+    except Exception as e:
+        raise MlflowException(
+            "The provided model configuration cannot be created as a Pipeline. "
+            "Please verify that all required and compatible components are "
+            "specified with the correct keys.",
+            error_code=INVALID_PARAMETER_VALUE,
+        ) from e
+
+
+def _record_pipeline_components(pipeline) -> Optional[Dict[str, Any]]:
+    """
+    Utility for recording which components are present in either the generated pipeline iff the
+    supplied save object is not a pipeline or the components of the supplied pipeline object.
+    """
+    components_conf = {}
+    components = []
+    if pipeline.feature_extractor:
+        feature_extractor = pipeline.feature_extractor
+        components_conf.update(
+            {_FEATURE_EXTRACTOR_TYPE_KEY: _get_instance_type(feature_extractor, False)}
+        )
+        components.append(_FEATURE_EXTRACTOR_KEY)
+    if pipeline.tokenizer:
+        tokenizer = pipeline.tokenizer
+        components_conf.update({_TOKENIZER_TYPE_KEY: _get_instance_type(tokenizer, False)})
+        components.append(_TOKENIZER_KEY)
+    if pipeline.image_processor:
```

Comment:
Some pipelines don't seem to have this attribute. I'm guessing other pipelines might not have `feature_extractor` or `tokenizer` attributes. Can we add `hasattr` checks here?

For reference, I ran the following code:

```
from transformers import AutoTokenizer, AutoModelWithLMHead, pipeline

tokenizer = AutoTokenizer.from_pretrained('microsoft/DialoGPT-small')
model = AutoModelWithLMHead.from_pretrained('satvikag/chatbot')

import mlflow.transformers

mlflow.transformers.log_model(
   pipeline(task="conversational", model=model, tokenizer=tokenizer),
   artifact_path="model",
)
```

Which yielded the following error:

```
2023-03-27 21:38:05.408850: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/transformers/models/auto/modeling_auto.py:998: FutureWarning: The class `AutoModelWithLMHead` is deprecated and will be removed in a future version. Please use `AutoModelForCausalLM` for causal language models, `AutoModelForMaskedLM` for masked language models and `AutoModelForSeq2SeqLM` for encoder-decoder models.
  warnings.warn(
Traceback (most recent call last):
  File "chat.py", line 35, in <module>
    mlflow.transformers.log_model(
  File "/Users/corey.zumar/mlflow/mlflow/transformers.py", line 467, in log_model
    return Model.log(
  File "/Users/corey.zumar/mlflow/mlflow/models/model.py", line 547, in log
    flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
  File "/Users/corey.zumar/mlflow/mlflow/transformers.py", line 270, in save_model
    components = _record_pipeline_components(built_pipeline)
  File "/Users/corey.zumar/mlflow/mlflow/transformers.py", line 652, in _record_pipeline_components
    if pipeline.image_processor:
AttributeError: 'ConversationalPipeline' object has no attribute 'image_processor'
```

Can we add test coverage for this as well?

Reply:
Thank you for finding this. I'll add guards around each of those component extraction checks and add another model type to validate this in tests.